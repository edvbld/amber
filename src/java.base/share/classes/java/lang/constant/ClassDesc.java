/*
 * Copyright (c) 2018, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */
package java.lang.constant;

import java.lang.invoke.MethodHandles;
import java.lang.invoke.TypeDescriptor;
import java.util.stream.Stream;

import sun.invoke.util.Wrapper;

import static java.lang.constant.ConstantUtils.binaryToInternal;
import static java.lang.constant.ConstantUtils.dropLastChar;
import static java.lang.constant.ConstantUtils.internalToBinary;
import static java.lang.constant.ConstantUtils.validateMemberName;
import static java.util.Objects.requireNonNull;
import static java.util.stream.Collectors.joining;

/**
 * A <a href="package-summary.html#nominal">nominal descriptor</a> for a
 * {@link Class} constant.
 *
 * <p>For common system types, including all the primitive types, there are
 * predefined {@linkplain ClassDesc} constants in {@link ConstantDescs}.  To create
 * a {@linkplain ClassDesc} for a class or interface type, use {@link #of} or
 * {@link #ofDescriptor(String)}; to create a {@linkplain ClassDesc} for an array
 * type, use {@link #ofDescriptor(String)}, or first obtain a
 * {@linkplain ClassDesc} for the component type and then call the {@link #arrayType()}
 * or {@link #arrayType(int)} methods.
 *
 * @apiNote In the future, if the Java language permits, {@linkplain ClassDesc}
 * may become a {@code sealed} interface, which would prohibit subclassing except
 * by explicitly permitted types.  Non-platform classes should not implement
 * {@linkplain ClassDesc} directly.
 *
 * @see ConstantDescs
 */
public interface ClassDesc
        extends ConstantDesc<Class<?>>,
                Constable<ConstantDesc<Class<?>>>,
                TypeDescriptor.OfField<ClassDesc> {

    /**
     * Create a {@linkplain ClassDesc} given the name of a class or interface
     * type, such as {@code "java.lang.String"}.  (To create a descriptor for an
     * array type, either use {@link #ofDescriptor(String)}
     * or {@link #arrayType()}; to create a descriptor for a primitive type, use
     * {@link #ofDescriptor(String)} or use the predefined constants in
     * {@link ConstantDescs}).
     *
     * @param name the fully qualified (dot-separated) binary class name
     * @return a {@linkplain ClassDesc} describing the desired class
     * @throws NullPointerException if any argument is {@code null}
     * @throws IllegalArgumentException if the name string is not in the
     * correct format
     */
    static ClassDesc of(String name) {
        ConstantUtils.validateBinaryClassName(requireNonNull(name));
        return ClassDesc.ofDescriptor("L" + binaryToInternal(name) + ";");
    }

    /**
     * Create a {@linkplain ClassDesc} given a package name and an unqualified
     * class name.
     *
     * @param packageName the package name (dot-separated)
     * @param className the unqualified class name
     * @return a {@linkplain ClassDesc} describing the desired class
     * @throws NullPointerException if any argument is {@code null}
     * @throws IllegalArgumentException if the package name or class name are
     * not in the correct format
     */
    static ClassDesc of(String packageName, String className) {
        ConstantUtils.validateBinaryClassName(requireNonNull(packageName));
        validateMemberName(requireNonNull(className));
        return ofDescriptor(String.format("L%s%s%s;",
                                          binaryToInternal(packageName),
                                          (packageName.length() > 0 ? "/" : ""),
                                          className));
    }

    /**
     * Create a {@linkplain ClassDesc} given a descriptor string.
     *
     * @param descriptor a field descriptor string, as per JVMS 4.3.2
     * @return a {@linkplain ClassDesc} describing the desired class
     * @throws NullPointerException if any argument is {@code null}
     * @throws IllegalArgumentException if the name string is not in the
     * correct format
     * @jvms 4.3.2 Field Descriptors
     */
    static ClassDesc ofDescriptor(String descriptor) {
        requireNonNull(descriptor);
        return (descriptor.length() == 1)
               ? new PrimitiveClassDescImpl(descriptor)
               : new ReferenceClassDescImpl(descriptor);
    }

    /**
     * Create a {@linkplain ClassDesc} for an array type whose component type
     * is described by this {@linkplain ClassDesc}.
     *
     * @return a {@linkplain ClassDesc} describing the array type
     */
    default ClassDesc arrayType() {
        return arrayType(1);
    }

    /**
     * Create a {@linkplain ClassDesc} for an array type of the specified rank,
     * whose component type is described by this {@linkplain ClassDesc}.
     *
     * @param rank the rank of the array
     * @return a {@linkplain ClassDesc} describing the array type
     * @throws IllegalArgumentException if the rank is zero or negative
     */
    default ClassDesc arrayType(int rank) {
        if (rank <= 0)
            throw new IllegalArgumentException("rank: " + rank);
        return ClassDesc.ofDescriptor("[".repeat(rank) + descriptorString());
    }

    /**
     * Create a {@linkplain ClassDesc} for an inner class of the class or
     * interface type described by this {@linkplain ClassDesc}.
     *
     * @param innerName the unqualified name of the inner class
     * @return a {@linkplain ClassDesc} describing the inner class
     * @throws NullPointerException if any argument is {@code null}
     * @throws IllegalStateException if this {@linkplain ClassDesc} does not
     * describe a class or interface type
     */
    default ClassDesc inner(String innerName) {
        validateMemberName(innerName);
        if (!isClassOrInterface())
            throw new IllegalStateException("Outer class is not a class or interface type");
        return ClassDesc.ofDescriptor(String.format("%s$%s;", dropLastChar(descriptorString()), innerName));
    }

    /**
     * Create a {@linkplain ClassDesc} for an inner class of the class or
     * interface type described by this {@linkplain ClassDesc}.
     *
     * @param firstInnerName the unqualified name of the first level of inner class
     * @param moreInnerNames the unqualified name(s) of the remaining levels of
     *                       inner class
     * @return a {@linkplain ClassDesc} describing the inner class
     * @throws NullPointerException if any argument is {@code null}
     * @throws IllegalStateException if this {@linkplain ClassDesc} does not
     * describe a class or interface type
     */
    default ClassDesc inner(String firstInnerName, String... moreInnerNames) {
        if (!isClassOrInterface())
            throw new IllegalStateException("Outer class is not a class or interface type");
        return moreInnerNames.length == 0
               ? inner(firstInnerName)
               : inner(firstInnerName + Stream.of(moreInnerNames).collect(joining("$", "$", "")));
    }

    /**
     * Returns whether this {@linkplain ClassDesc} describes an array type.
     *
     * @return whether this {@linkplain ClassDesc} describes an array type
     */
    default boolean isArray() {
        return descriptorString().startsWith("[");
    }

    /**
     * Returns whether this {@linkplain ClassDesc} describes a primitive type.
     *
     * @return whether this {@linkplain ClassDesc} describes a primitive type
     */
    default boolean isPrimitive() {
        return descriptorString().length() == 1;
    }

    /**
     * Returns whether this {@linkplain ClassDesc} describes a class or interface type.
     *
     * @return whether this {@linkplain ClassDesc} describes a class or interface type
     */
    default boolean isClassOrInterface() {
        return descriptorString().startsWith("L");
    }

    /**
     * Returns the component type of this {@linkplain ClassDesc}, if it describes
     * an array type, or {@code null} otherwise.
     *
     * @return a {@linkplain ClassDesc} describing the component type, or {@code null}
     * if this descriptor does not describe an array type
     */
    default ClassDesc componentType() {
        return isArray() ? ClassDesc.ofDescriptor(descriptorString().substring(1)) : null;
    }

    /**
     * Returns the package name of this {@linkplain ClassDesc}, if it describes
     * a class or interface type.
     *
     * @return the package name, or the empty string if the class is in the
     * default package
     * @throws IllegalStateException if this {@linkplain ClassDesc} does not
     * describe a class or interface type
     */
    default String packageName() {
        if (!isClassOrInterface())
            throw new IllegalStateException("not a class or interface");
        String className = internalToBinary(ConstantUtils.dropFirstAndLastChar(descriptorString()));
        int index = className.lastIndexOf('.');
        return (index == -1) ? "" : className.substring(0, index);
    }

    /**
     * Returns a human-readable name for the type described by this descriptor.
     *
     * @return the human-readable name
     */
    default String displayName() {
        if (isPrimitive())
            return Wrapper.forBasicType(descriptorString().charAt(0)).primitiveSimpleName();
        else if (isClassOrInterface()) {
            return descriptorString().substring(Math.max(1, descriptorString().lastIndexOf('/') + 1),
                                                descriptorString().length() - 1);
        }
        else if (isArray()) {
            int depth = ConstantUtils.arrayDepth(descriptorString());
            ClassDesc c = this;
            for (int i=0; i<depth; i++)
                c = c.componentType();
            return c.displayName() + "[]".repeat(depth);
        }
        else
            throw new IllegalStateException(descriptorString());
    }

    /**
     * Constant bootstrap method for representing a {@linkplain ClassDesc} in
     * the constant pool of a classfile.
     *
     * @param lookup ignored
     * @param name ignored
     * @param clazz ignored
     * @param descriptor a field descriptor string for the class, as per JVMS 4.3.2
     * @return the {@linkplain ClassDesc}
     * @jvms 4.3.2 Field Descriptors
     */
    public static ClassDesc constantBootstrap(MethodHandles.Lookup lookup, String name, Class<ClassDesc> clazz,
                                              String descriptor) {
        return ClassDesc.ofDescriptor(descriptor);
    }

    /**
     * Return a field type descriptor string for this type, as per JVMS 4.3.2
     *
     * @return the descriptor string
     * @jvms 4.3.2 Field Descriptors
     */
    String descriptorString();
}
