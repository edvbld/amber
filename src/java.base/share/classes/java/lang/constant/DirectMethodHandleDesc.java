/*
 * Copyright (c) 2018, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */
package java.lang.constant;

import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandleInfo;
import java.util.OptionalInt;
import java.util.stream.Stream;

import jdk.internal.vm.annotation.Stable;

import static java.lang.invoke.MethodHandleInfo.REF_getField;
import static java.lang.invoke.MethodHandleInfo.REF_getStatic;
import static java.lang.invoke.MethodHandleInfo.REF_invokeInterface;
import static java.lang.invoke.MethodHandleInfo.REF_invokeSpecial;
import static java.lang.invoke.MethodHandleInfo.REF_invokeStatic;
import static java.lang.invoke.MethodHandleInfo.REF_invokeVirtual;
import static java.lang.invoke.MethodHandleInfo.REF_newInvokeSpecial;
import static java.lang.invoke.MethodHandleInfo.REF_putField;
import static java.lang.invoke.MethodHandleInfo.REF_putStatic;

/**
 * A <a href="package-summary.html#nominal">nominal descriptor</a> for a direct
 * {@link MethodHandle}.  A {@linkplain DirectMethodHandleDescImpl} corresponds to
 * a {@code Constant_MethodHandle_info} entry in the constant pool of a classfile.
 *
 * @apiNote In the future, if the Java language permits, {@linkplain DirectMethodHandleDesc}
 * may become a {@code sealed} interface, which would prohibit subclassing except
 * by explicitly permitted types.  Non-platform classes should not implement
 * {@linkplain DirectMethodHandleDesc} directly.
 */
public interface DirectMethodHandleDesc extends MethodHandleDesc {
    /**
     * Kinds of method handles that can be described with {@linkplain DirectMethodHandleDesc}.
     */
    enum Kind {
        /** A method handle for a method invoked as with {@code invokestatic} */
        STATIC(REF_invokeStatic),
        /** A method handle for a method invoked as with {@code invokestatic} */
        INTERFACE_STATIC(REF_invokeStatic, true),
        /** A method handle for a method invoked as with {@code invokevirtual} */
        VIRTUAL(REF_invokeVirtual),
        /** A method handle for a method invoked as with {@code invokeinterface} */
        INTERFACE_VIRTUAL(REF_invokeInterface, true),
        /** A method handle for a method invoked as with {@code invokespecial} */
        SPECIAL(REF_invokeSpecial),
        /** A method handle for an interface method invoked as with {@code invokespecial} */
        INTERFACE_SPECIAL(REF_invokeSpecial, true),
        /** A method handle for a constructor */
        CONSTRUCTOR(REF_newInvokeSpecial),
        /** A method handle for a read accessor for an instance field  */
        GETTER(REF_getField),
        /** A method handle for a write accessor for an instance field  */
        SETTER(REF_putField),
        /** A method handle for a read accessor for a static field  */
        STATIC_GETTER(REF_getStatic),
        /** A method handle for a write accessor for a static field  */
        STATIC_SETTER(REF_putStatic);

        /** The corresponding {@code refKind} value for this kind of method handle,
         * as defined by {@link MethodHandleInfo}
         */
        public final int refKind;

        /** Is this an interface
         */
        public final boolean isInterface;
        Kind(int refKind) {
            this(refKind, false);
        }

        Kind(int refKind, boolean isInterface) { this.refKind = refKind; this.isInterface = isInterface; }

        /**
         * Find the enumeration member with the given {@code refKind} field.
         * Behaves as if {@code valueOf(refKind, false)}.  As a special case,
         * if {@code refKind} is {@code REF_invokeInterface} (9) then the
         * {@code isInterface} field will be true.
         *
         * @param refKind refKind of desired member
         * @return the matching enumeration member
         * @throws IllegalArgumentException if there is no such member
         */
        public static Kind valueOf(int refKind) {
            return valueOf(refKind, false);
        }

        /**
         * Find the enumeration member with the given {@code refKind} and
         * {@code isInterface} fields. If {@code isInterface} is true and there
         * is no such enumeration member, return the member, if any, with the
         * same {@code refKind} and a false {@code isInterface} field. If
         * {@code isInterface} is true but there is no such enumeration member,
         * then the result of {@code valueOf(refKind, false)} is returned.  As
         * a special case, if {@code refKind} is {@code REF_invokeVirtual} (5) and
         * {@code isInterface} is true, then the result of
         * {@code valueOf(REF_invokeInterface, false)} is returned, and if
         * {@code isInterface} is false and {@code refKind} is {@code REF_invokeInterface},
         * {@code INTERFACE_VIRTUAL} is returned.
         *
         * @param refKind refKind of desired member
         * @param isInterface whether desired member is for interface methods
         * @return the matching enumeration member
         * @throws IllegalArgumentException if there is no such member
         */
        public static Kind valueOf(int refKind, boolean isInterface) {
            int i = tableIndex(refKind, isInterface);
            if (i >= 0 && i < TABLE.length) {
                Kind kind = TABLE[i];
                if (kind.refKind == refKind && kind.isInterface == isInterface) {
                    return kind;
                }
            }
            if (isInterface)
                return valueOf(refKind);
            else if (refKind == REF_invokeInterface)
                return INTERFACE_VIRTUAL;
            else
                throw new IllegalArgumentException(String.format("refKind=%d", refKind));
        }

        private static int tableIndex(int refKind, boolean isInterface) {
            if (refKind < 0)  return refKind;
            return (refKind * 2) + (isInterface ? 1 : 0);
        }

        private static final @Stable Kind[] TABLE;

        static {
            // Pack the static table.
            int max = 0;
            for (Kind k : values())
                max = Math.max(max, tableIndex(k.refKind, k.isInterface));

            TABLE = new Kind[max+1];
            for (Kind kind : values()) {
                int i = tableIndex(kind.refKind, kind.isInterface);
                if (i >= TABLE.length || TABLE[i] != null)
                    throw new AssertionError("TABLE entry for " + kind);
                TABLE[i] = kind;
            }

            // Pack in some aliases also.
            int ii = tableIndex(REF_invokeInterface, false);
            if (TABLE[ii] != null)
                throw new AssertionError("TABLE entry for (invokeInterface, false) used by " + TABLE[ii]);
            TABLE[ii] = INTERFACE_VIRTUAL;

            for (Kind kind : values()) {
                if (!kind.isInterface) {
                    // Add extra cache entry to alias the isInterface case.
                    // For example, (REF_getStatic, X) will produce STATIC_GETTER
                    // for either truth value of X.
                    int i = tableIndex(kind.refKind, true);
                    if (TABLE[i] == null) {
                        // There is not a specific Kind for interfaces
                        if (kind == VIRTUAL)  kind = INTERFACE_VIRTUAL;
                        if (TABLE[i] == null)  TABLE[i] = kind;
                    }
                }
            }
        }
    }

    /**
     * Return the {@code kind} of the method handle described by this nominal
     * descriptor.
     *
     * @return the {@link Kind}
     */
    Kind kind();

    /**
     * Return the {@code refKind} of the method handle described by this nominal
     * reference, as defined by {@link MethodHandleInfo}.
     *
     * @return the reference kind
     */
    int refKind();

    /**
     * Indicates if the method is declared by an interface
     *
     * @return true if the method is declared by an interface
     */
    boolean isOwnerInterface();

    /**
     * Return a {@link ClassDesc} describing the class declaring the
     * method or field described by this nominal descriptor.
     *
     * @return the class declaring the method or field
     */
    ClassDesc owner();

    /**
     * Return the name of the method or field described by this nominal descriptor.
     *
     * @return the name of the method or field
     */
    String methodName();

    /**
     * Return a {@link MethodTypeDesc} describing the invocation type of the
     * method handle described by this nominal descriptor
     *
     * @return the method type
     */
    MethodTypeDesc methodType();
}
