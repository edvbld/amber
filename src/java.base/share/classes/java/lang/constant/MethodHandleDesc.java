/*
 * Copyright (c) 2018, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */
package java.lang.constant;

import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodType;

import static java.lang.constant.ConstantDescs.CD_void;
import static java.lang.constant.DirectMethodHandleDesc.Kind.CONSTRUCTOR;

/**
 * A <a href="package-summary.html#nominal">nominal descriptor</a> for a
 * {@link MethodHandle} constant.
 *
 * @apiNote In the future, if the Java language permits, {@linkplain MethodHandleDesc}
 * may become a {@code sealed} interface, which would prohibit subclassing except
 * by explicitly permitted types.  Non-platform classes should not implement
 * {@linkplain MethodHandleDesc} directly.
 */
public interface MethodHandleDesc
        extends ConstantDesc<MethodHandle> {

    /**
     * Create a {@linkplain MethodHandleDesc} corresponding to an invocation of a
     * declared method, invocation of a constructor, or access to a field.
     *
     * <p>If {@code kind} is {@code CONSTRUCTOR}, the name is ignored and the return
     * type of the invocation type must be {@code void}.  If {@code kind} corresponds
     * to a field access, the invocation type must be consistent with that kind
     * of field access and the type of the field; instance field accessors must
     * take a leading receiver parameter, getters must return the type of the
     * field, setters must take a new value for the field and return {@code void}.
     *
     * <p>For constructor and field access, the methods {@link #ofField(DirectMethodHandleDesc.Kind, ClassDesc, String, ClassDesc)}
     * and {@link #ofConstructor(ClassDesc, ClassDesc...)} may be more convenient.
     *
     * @param kind The kind of method handle to be described
     * @param clazz a {@link ClassDesc} describing the class containing the
     *              method, constructor, or field
     * @param name the name of the method or field (ignored if {@code kind} is
     * {@code CONSTRUCTOR}), as per JVMS 4.2.2
     * @param type a {@link MethodTypeDesc} describing the invocation type of
     *             the method handle
     * @return the {@linkplain MethodHandleDesc}
     * @throws NullPointerException if any non-ignored arguments are null
     * @throws IllegalArgumentException if the {@code name} has the incorrect
     * format
     * @jvms 4.2.2 Unqualified Names
     */
    static DirectMethodHandleDesc of(DirectMethodHandleDesc.Kind kind,
                                     ClassDesc clazz,
                                     String name,
                                     MethodTypeDesc type) {
        return new DirectMethodHandleDescImpl(kind, clazz, name, type);
    }

    /**
     * Create a {@linkplain MethodHandleDesc} corresponding to an invocation of a
     * declared method, invocation of a constructor, or access to a field.
     *
     * <p>If {@code kind} is {@code CONSTRUCTOR}, the name is ignored and the return
     * type of the invocation type must be {@code void}.  If {@code kind} corresponds
     * to a field access, the invocation type must be consistent with that kind
     * of field access and the type of the field; instance field accessors must
     * take a leading receiver parameter, getters must return the type of the
     * field, setters must take a new value for the field and return {@code void}.
     * The method {@link #ofField(DirectMethodHandleDesc.Kind, ClassDesc, String, ClassDesc)} will construct
     * the appropriate invocation given the type of the field.
     *
     * @param kind The kind of method handle to be described
     * @param clazz a {@link ClassDesc} describing the class containing the
     *              method, constructor, or field
     * @param name the name of the method or field (ignored if {@code kind} is
     * {@code CONSTRUCTOR}), as per JVMS 4.2.2
     * @param descriptorString a method descriptor string for the invocation type
     * of the method handle, as per JVMS 4.3.3
     * @return the {@linkplain MethodHandleDesc}
     * @throws NullPointerException if any of the non-ignored arguments are null
     * @jvms 4.2.2 Unqualified Names
     * @jvms 4.3.3 Method Descriptors
     */
    static DirectMethodHandleDesc of(DirectMethodHandleDesc.Kind kind,
                                     ClassDesc clazz,
                                     String name,
                                     String descriptorString) {
        return of(kind, clazz, name, MethodTypeDesc.ofDescriptor(descriptorString));
    }

    /**
     * Create a {@linkplain MethodHandleDesc} corresponding to an invocation of a
     * declared method, invocation of a constructor, or access to a field.
     *
     * <p>If {@code kind} is {@code CONSTRUCTOR}, the name is ignored and the return
     * type of the invocation type must be {@code void}.  If {@code kind} corresponds
     * to a field access, the invocation type must be consistent with that kind
     * of field access and the type of the field; instance field accessors must
     * take a leading receiver parameter, getters must return the type of the
     * field, setters must take a new value for the field and return {@code void}.
     * The method {@link #ofField(DirectMethodHandleDesc.Kind, ClassDesc, String, ClassDesc)} will construct
     * the appropriate invocation given the type of the field.
     *
     * @param kind The kind of method handle to be described
     * @param clazz a {@link ClassDesc} describing the class containing the
     *              method, constructor, or field
     * @param name the name of the method or field (ignored if {@code kind} is
     * {@code CONSTRUCTOR}), as per JVMS 4.2.2
     * @param returnType a {@link ClassDesc} describing the return type of the
     *                   method handle
     * @param paramTypes {@link ClassDesc}s describing the parameter types of
     *                                    the method handle
     * @return the {@linkplain MethodHandleDesc}
     * @throws NullPointerException if any of the non-ignored arguments are null
     * @jvms 4.2.2 Unqualified Names
     */
    static DirectMethodHandleDesc of(DirectMethodHandleDesc.Kind kind,
                                     ClassDesc clazz,
                                     String name,
                                     ClassDesc returnType,
                                     ClassDesc... paramTypes) {
        return of(kind, clazz, name, MethodTypeDesc.of(returnType, paramTypes));
    }

    /**
     * Create a {@linkplain MethodHandleDesc} corresponding to a method handle
     * that accesses a field.
     *
     * @param kind the kind of the method handle to be described; must be one of {@code GETTER},
     *             {@code SETTER}, {@code STATIC_GETTER}, or {@code STATIC_SETTER}
     * @param clazz a {@link ClassDesc} describing the class containing the
     *              method, constructor, or field
     * @param fieldName the name of the field, as per JVMS 4.2.2
     * @param fieldType a {@link ClassDesc} describing the type of the field
     * @return the {@linkplain MethodHandleDesc}
     * @throws NullPointerException if any of the arguments are null
     * @jvms 4.2.2 Unqualified Names
     */
    static DirectMethodHandleDesc ofField(DirectMethodHandleDesc.Kind kind,
                                          ClassDesc clazz,
                                          String fieldName,
                                          ClassDesc fieldType) {
        MethodTypeDesc mtr;
        switch (kind) {
            case GETTER: mtr = MethodTypeDesc.of(fieldType, clazz); break;
            case SETTER: mtr = MethodTypeDesc.of(CD_void, clazz, fieldType); break;
            case STATIC_GETTER: mtr = MethodTypeDesc.of(fieldType); break;
            case STATIC_SETTER: mtr = MethodTypeDesc.of(CD_void, fieldType); break;
            default:
                throw new IllegalArgumentException(kind.toString());
        }
        return MethodHandleDesc.of(kind, clazz, fieldName, mtr);
    }

    /**
     * Return a {@linkplain MethodHandleDesc} corresponding to invocation of a constructor
     *
     * @param clazz a {@link ClassDesc} describing the class containing the
     *              method, constructor, or field
     * @param paramTypes {@link ClassDesc}s describing the parameter types of
     *                   the constructor
     * @return the {@linkplain MethodHandleDesc}
     * @throws NullPointerException if any of the arguments are null
     */
    static DirectMethodHandleDesc ofConstructor(ClassDesc clazz,
                                                ClassDesc... paramTypes) {
        return MethodHandleDesc.of(CONSTRUCTOR, clazz, ConstantDescs.DEFAULT_NAME,
                                   MethodTypeDesc.of(CD_void, paramTypes));
    }

    /**
     * Return a {@linkplain MethodHandleDesc} that describes this method handle
     * adapted to a different type, as if by {@link MethodHandle#asType(MethodType)}.
     *
     * @param type a {@link MethodHandleDesc} describing the new method type
     * @return a {@linkplain MethodHandleDesc} for the adapted method handle
     */
    default MethodHandleDesc asType(MethodTypeDesc type) {
        return (methodType().equals(type)) ? this : new AsTypeMethodHandleDesc(this, type);
    }

    /**
     * Return a {@link MethodTypeDesc} describing the type of the method handle
     * described by this nominal descriptor
     *
     * @return a {@linkplain MethodHandleDesc} describing the method handle type
     */
    MethodTypeDesc methodType();
}
