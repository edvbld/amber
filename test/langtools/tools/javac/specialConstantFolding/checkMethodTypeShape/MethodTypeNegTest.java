/*
 * @test /nodynamiccopyright/
 * @summary check that method type's shape matches the reference kind
 * @compile/fail/ref=MethodTypeNegTest.out -XDdoConstantFold -Werror -XDrawDiagnostics MethodTypeNegTest.java
 */

import java.lang.invoke.constant.ClassDesc;
import java.lang.invoke.constant.ConstantDescs;
import java.lang.invoke.constant.MethodHandleDesc;


public class MethodTypeNegTest {
    private static final ClassDesc CR_THIS = ClassDesc.of("MethodTypeNegTest");

    void test() {
        // for setters result must be void...
        MethodHandleDesc r1 = MethodHandleDesc.of(MethodHandleDesc.Kind.SETTER, CR_THIS, "nonexistent", ConstantDescs.CR_int);
        MethodHandleDesc r2 = MethodHandleDesc.of(MethodHandleDesc.Kind.STATIC_SETTER, CR_THIS, "nonexistent", ConstantDescs.CR_int);

        // ...and static setters should have only one argument and instance setters two
        MethodHandleDesc r3 = MethodHandleDesc.of(MethodHandleDesc.Kind.SETTER, CR_THIS, "nonexistent", ConstantDescs.CR_void, ConstantDescs.CR_int);
        MethodHandleDesc r4 = MethodHandleDesc.of(MethodHandleDesc.Kind.STATIC_SETTER, CR_THIS, "nonexistent", ConstantDescs.CR_void, ConstantDescs.CR_int, ConstantDescs.CR_int);

        // for constructors result type must be void
        MethodHandleDesc r5 = MethodHandleDesc.of(MethodHandleDesc.Kind.CONSTRUCTOR, CR_THIS, "", ConstantDescs.CR_int);

        // for getters result must be different from void...
        MethodHandleDesc r6 = MethodHandleDesc.of(MethodHandleDesc.Kind.GETTER, CR_THIS, "nonexistent", ConstantDescs.CR_void);
        MethodHandleDesc r7 = MethodHandleDesc.of(MethodHandleDesc.Kind.STATIC_GETTER, CR_THIS, "nonexistent", ConstantDescs.CR_void);

        // ...and instance setters should have only one argument, static ones none
        MethodHandleDesc r8 = MethodHandleDesc.of(MethodHandleDesc.Kind.GETTER, CR_THIS, "nonexistent", ConstantDescs.CR_int, ConstantDescs.CR_int, ConstantDescs.CR_int);
        MethodHandleDesc r9 = MethodHandleDesc.of(MethodHandleDesc.Kind.STATIC_GETTER, CR_THIS, "nonexistent", ConstantDescs.CR_int, ConstantDescs.CR_int, ConstantDescs.CR_int);

        // no argument can be void
        MethodHandleDesc r10 = MethodHandleDesc.of(MethodHandleDesc.Kind.VIRTUAL, CR_THIS, "nonexistent", ConstantDescs.CR_int, ConstantDescs.CR_void);
    }
}
