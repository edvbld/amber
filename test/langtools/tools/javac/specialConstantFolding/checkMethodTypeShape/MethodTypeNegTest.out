MethodTypeNegTest.java:20:48: compiler.err.reflective.error: of, java.lang.invoke.constant.MethodHandleRef, Expected type of (RT)V for getter, found MethodTypeRef[()int]
MethodTypeNegTest.java:21:48: compiler.err.reflective.error: of, java.lang.invoke.constant.MethodHandleRef, Expected type of (T)V for getter, found MethodTypeRef[()int]
MethodTypeNegTest.java:24:48: compiler.err.reflective.error: of, java.lang.invoke.constant.MethodHandleRef, Expected type of (RT)V for getter, found MethodTypeRef[(int)void]
MethodTypeNegTest.java:25:48: compiler.err.reflective.error: of, java.lang.invoke.constant.MethodHandleRef, Expected type of (T)V for getter, found MethodTypeRef[(int,int)void]
MethodTypeNegTest.java:28:48: compiler.err.reflective.error: of, java.lang.invoke.constant.MethodHandleRef, Expected type of (T*)V for constructor, found MethodTypeRef[()int]
MethodTypeNegTest.java:31:48: compiler.err.reflective.error: of, java.lang.invoke.constant.MethodHandleRef, Expected type of (R)T for getter, found MethodTypeRef[()void]
MethodTypeNegTest.java:32:48: compiler.err.reflective.error: of, java.lang.invoke.constant.MethodHandleRef, Expected type of ()T for getter, found MethodTypeRef[()void]
MethodTypeNegTest.java:35:48: compiler.err.reflective.error: of, java.lang.invoke.constant.MethodHandleRef, Expected type of (R)T for getter, found MethodTypeRef[(int,int)int]
MethodTypeNegTest.java:36:48: compiler.err.reflective.error: of, java.lang.invoke.constant.MethodHandleRef, Expected type of ()T for getter, found MethodTypeRef[(int,int)int]
MethodTypeNegTest.java:39:49: compiler.err.reflective.error: of, java.lang.invoke.constant.MethodHandleRef, Void parameters not permitted
10 errors
